k2/csrc/tensor.cu

```diff
@@ -131,9 +132,57 @@ void Tensor::Init(ContextPtr c) {
 }
 
 Tensor ToContiguous(const Tensor &tensor) {
-  // TODO(haowen): implement
-  K2_LOG(FATAL) << "Not implemented";
-  return tensor;
+  if (tensor.IsContiguous()) return tensor;
+  std::vector<int32_t> dims = tensor.Dims();
+  std::vector<int32_t> strides = tensor.Strides();
+  int32_t num_dims = static_cast<int32_t>(dims.size());
+
+  std::size_t num_dims_bytes = num_dims * sizeof(int32_t);
+
+  RegionPtr dims_ptr = NewRegion(tensor.Context(), num_dims_bytes);
+  RegionPtr strides_ptr = NewRegion(tensor.Context(), num_dims_bytes);
+
+  auto cpu_context = tensor.Context()->GetCpuContext();
+  auto kind = GetMemoryCopyKind(*cpu_context, *tensor.Context());
+  MemoryCopy(dims_ptr->data, static_cast<void *>(dims.data()), num_dims_bytes,
+             kind);
+  MemoryCopy(strides_ptr->data, static_cast<void *>(strides.data()),
+             num_dims_bytes, kind);
+
+  const int32_t *dims_data = dims_ptr->GetData<int32_t>();
+  const int32_t *strides_data = strides_ptr->GetData<int32_t>();
+
+  // turn a linear index of a multi-dimensional array
+  // first into a coordinate representation, and then to offset
+  // in number of elements via strides information.
+  auto to_offset = [=] __host__ __device__(int32_t i) -> int32_t {
+    int32_t offset = 0;
+    for (int32_t k = num_dims - 1; k > 0; --k) {
+      int32_t index = i % dims_data[k];
+      offset += index * strides_data[k];
+      i = (i - index) / dims_data[k];
+      if (i == 0) break;
```

Comment:
I dont see the need for this if-statement.  If can slow things down quite a bit.

Reply:
Yes, you are right. I just wanted it to break the loop earlier. I will fix it.

Reply:
Removed.